possible Gossip deadlock in TestRangeSplitsWithWritePressure

The following test appears to have failed:

[#915](https://circleci.com/gh/cockroachdb/cockroach/915):

```
I0326 05:44:16.705357      69 multiraft.go:631] HardState updated: {Term:6 Vote:257 Commit:117 XXX_unrecognized:[]}
I0326 05:44:16.705514      69 multiraft.go:634] New Entry[0]: 6/117 EntryNormal 00000000000000007078d73cd7485ab0: raft_id:2 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:765 > cmd_id:<wall_time:0 random:0 > key:"sVJgQbWLmM" user:"root" rep
I0326 05:44:16.705660      69 multiraft.go:637] Committed Entry[0]: 6/117 EntryNormal 00000000000000007078d73cd7485ab0: raft_id:2 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:765 > cmd_id:<wall_time:0 random:0 > key:"sVJgQbWLmM" user:"root" rep
I0326 05:44:16.706278      69 queue.go:207] processing range range=1 (""-"IZNbVZhCts") from split queue...
I0326 05:44:16.707374      69 range.go:1858] initiating a split of range 1 "\"\""-"\"IZNbVZhCts\"" at key "\"FnxZTCgeqD\""
panic: test timed out after 30s

goroutine 729 [running]:
testing.func·008()
	/usr/src/go/src/testing/testing.go:681 +0x12f
created by time.goFunc
	/usr/src/go/src/time/sleep.go:129 +0x4b

goroutine 1 [chan receive]:
testing.RunTests(0xfdd788, 0x145a340, 0x2b, 0x2b, 0xc208029a01)
	/usr/src/go/src/testing/testing.go:556 +0xad6
--
	/go/src/github.com/cockroachdb/cockroach/storage/split_queue.go:117 +0xaa6
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop(0xc208182540, 0xc2080bce80, 0xc20821e0c0)
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:208 +0x55a
created by github.com/cockroachdb/cockroach/storage.(*baseQueue).Start
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:132 +0x5d
FAIL	github.com/cockroachdb/cockroach/kv	30.018s
=== RUN TestHeartbeatSingleGroup
I0326 05:44:16.717716      78 multiraft.go:408] node 1 starting
I0326 05:44:16.717848      78 multiraft.go:408] node 2 starting
I0326 05:44:16.717996      78 raft.go:315] raft: 1 became follower at term 5
I0326 05:44:16.718045      78 raft.go:134] raft: newRaft 1 [peers: [1,2], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0326 05:44:16.718148      78 raft.go:315] raft: 2 became follower at term 5
I0326 05:44:16.718183      78 raft.go:134] raft: newRaft 2 [peers: [1,2], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0326 05:44:16.718243      78 raft.go:389] raft: 1 is starting a new election at term 5
I0326 05:44:16.718290      78 raft.go:328] raft: 1 became candidate at term 6
I0326 05:44:16.718309      78 raft.go:372] raft: 1 received vote from 1 at term 6
I0326 05:44:16.705357      69 multiraft.go:631] HardState updated: {Term:6 Vote:257 Commit:117 XXX_unrecognized:[]}
I0326 05:44:16.705514      69 multiraft.go:634] New Entry[0]: 6/117 EntryNormal 00000000000000007078d73cd7485ab0: raft_id:2 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:765 > cmd_id:<wall_time:0 random:0 > key:"sVJgQbWLmM" user:"root" rep
I0326 05:44:16.705660      69 multiraft.go:637] Committed Entry[0]: 6/117 EntryNormal 00000000000000007078d73cd7485ab0: raft_id:2 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:765 > cmd_id:<wall_time:0 random:0 > key:"sVJgQbWLmM" user:"root" rep
I0326 05:44:16.706278      69 queue.go:207] processing range range=1 (""-"IZNbVZhCts") from split queue...
I0326 05:44:16.707374      69 range.go:1858] initiating a split of range 1 "\"\""-"\"IZNbVZhCts\"" at key "\"FnxZTCgeqD\""
panic: test timed out after 30s

goroutine 729 [running]:
testing.func·008()
	/usr/src/go/src/testing/testing.go:681 +0x12f
created by time.goFunc
	/usr/src/go/src/time/sleep.go:129 +0x4b

goroutine 1 [chan receive]:
testing.RunTests(0xfdd788, 0x145a340, 0x2b, 0x2b, 0xc208029a01)
	/usr/src/go/src/testing/testing.go:556 +0xad6
--
	/go/src/github.com/cockroachdb/cockroach/storage/split_queue.go:117 +0xaa6
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop(0xc208182540, 0xc2080bce80, 0xc20821e0c0)
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:208 +0x55a
created by github.com/cockroachdb/cockroach/storage.(*baseQueue).Start
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:132 +0x5d
FAIL	github.com/cockroachdb/cockroach/kv	30.018s
=== RUN TestHeartbeatSingleGroup
I0326 05:44:16.717716      78 multiraft.go:408] node 1 starting
I0326 05:44:16.717848      78 multiraft.go:408] node 2 starting
I0326 05:44:16.717996      78 raft.go:315] raft: 1 became follower at term 5
I0326 05:44:16.718045      78 raft.go:134] raft: newRaft 1 [peers: [1,2], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0326 05:44:16.718148      78 raft.go:315] raft: 2 became follower at term 5
I0326 05:44:16.718183      78 raft.go:134] raft: newRaft 2 [peers: [1,2], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0326 05:44:16.718243      78 raft.go:389] raft: 1 is starting a new election at term 5
I0326 05:44:16.718290      78 raft.go:328] raft: 1 became candidate at term 6
I0326 05:44:16.718309      78 raft.go:372] raft: 1 received vote from 1 at term 6

```
Please assign, take a look and update the issue accordingly.

--- a/kv/split_test.go
+++ b/kv/split_test.go
@@ -144,8 +144,8 @@ func TestRangeSplitsWithWritePressure(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer transport.Close()
 	defer db.Close()
+	defer transport.Close()
 	setTestRetryOptions()
 
 	// Rewrite a zone config with low max bytes.

--- a/kv/txn_coord_sender.go
+++ b/kv/txn_coord_sender.go
@@ -196,11 +196,15 @@ func (tc *TxnCoordSender) Send(call *client.Call) {
 // TxnCoordSender has been managing.
 func (tc *TxnCoordSender) Close() {
 	tc.Lock()
-	defer tc.Unlock()
 	for _, txn := range tc.txns {
 		close(txn.closer)
 	}
 	tc.txns = map[string]*txnMetadata{}
+	tc.Unlock()
+	// TODO(bdarnell): closing the wrapped sender may close the underlying
+	// Store, which may call back into TxnCoordSender and cause deadlocks if
+	// the lock is held. The shutdown process needs to be refactored to clean
+	// this up.
 	tc.wrapped.Close()
 }
 

--- a/storage/store.go
+++ b/storage/store.go
@@ -301,6 +301,10 @@ type Store struct {
 	verifyQueue *verifyQueue        // Checksum verification queue
 	scanner     *rangeScanner       // Range scanner
 	multiraft   *multiraft.MultiRaft
+	// scanStopper is used to stop background tasks that might be adding work to the
+	// store. stopper is used to stop the store itself. Deadlocks could result if
+	// the two used the same stopper.
+	scanStopper *util.Stopper
 	stopper     *util.Stopper
 
 	mu          sync.RWMutex     // Protects variables below...
@@ -324,6 +328,7 @@ func NewStore(clock *hlc.Clock, eng engine.Engine, db *client.KV, gossip *gossip
 		allocator:   &allocator{},
 		gossip:      gossip,
 		transport:   transport,
+		scanStopper: util.NewStopper(0),
 		stopper:     util.NewStopper(0),
 		ranges:      map[int64]*Range{},
 	}
@@ -339,9 +344,11 @@ func NewStore(clock *hlc.Clock, eng engine.Engine, db *client.KV, gossip *gossip
 	return s
 }
 
-// Stop calls Range.Stop() on all active ranges.
+// Stop shuts down the store.
 func (s *Store) Stop() {
-	// Stop the id allocator first since it may need the ranges to be up to finish.
+	// First, stop the scanner which might be generating new work in the background.
+	s.scanStopper.Stop()
+	// Second, stop the id allocator which may need the ranges to be up to finish.
 	if s.raftIDAlloc != nil {
 		s.raftIDAlloc.Stop()
 	}
@@ -449,7 +456,7 @@ func (s *Store) Start() error {
 	go s.processRaft()
 
 	// Start the scanner.
-	s.scanner.Start(s.clock, s.stopper)
+	s.scanner.Start(s.clock, s.scanStopper)
 
 	// Register callbacks for any changes to accounting and zone
 	// configurations; we split ranges along prefix boundaries.

